framework/src/headers/ip/v6/icmp/router_advertisement.rs

```diff
@@ -0,0 +1,339 @@
+        // Options are a variable length and will be manually parsed
+        16
+    }
+
+    #[inline]
+    fn size() -> usize {
+        // ICMPv6 Header is always 8 bytes so size = offset
+        16
+    }
+
+    #[inline]
+    fn payload_size(&self, hint: usize) -> usize {
+        // There is no payload size in the ICMPv6 header
+        hint - self.offset()
+    }
+
+    #[inline]
+    fn check_correct(&self, _prev: &T) -> bool {
+        true
+    }
+}
+
+impl<T> Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    #[inline]
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    #[inline]
+    pub fn msg_type(&self) -> Option<IcmpMessageType> {
+        self.icmp.msg_type()
+    }
+
+    #[inline]
+    pub fn code(&self) -> u8 {
+        self.icmp.code()
+    }
+
+    #[inline]
+    pub fn checksum(&self) -> u16 {
+        self.icmp.checksum()
+    }
+
+    #[inline]
+    pub fn current_hop_limit(&self) -> u8 {
+        self.current_hop_limit
+    }
+
+    #[inline]
+    pub fn reserved_flags(&self) -> u8 {
+        self.reserved_flags
+    }
+
+    #[inline]
+    pub fn managed_addr_cfg(&self) -> bool {
+        get_bit(self.reserved_flags, MANAGED_CFG_ADDR_POS)
+    }
+
+    #[inline]
+    pub fn other_cfg(&self) -> bool {
+        get_bit(self.reserved_flags, OTHER_CFG_POS)
+    }
+
+    #[inline]
+    pub fn router_lifetime(&self) -> u16 {
+        u16::from_be(self.router_lifetime)
+    }
+
+    #[inline]
+    pub fn reachable_time(&self) -> u32 {
+        u32::from_be(self.reachable_time)
+    }
+
+    #[inline]
+    pub fn retrans_timer(&self) -> u32 {
+        u32::from_be(self.retrans_timer)
+    }
+}
+
+impl<T> IPv6Optionable for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn parse_options(&self, payload_len: u16) -> HashMap<Icmpv6OptionType, Icmpv6Option> {
+        let mut options_map = HashMap::new();
+        unsafe {
+            let self_as_u8 = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset(); //track to make sure we don't go beyond v6 payload size
+
+            while payload_offset < (payload_len as usize) {
+                // start at beginning of the options
+                // the second byte is the header length in 8-octet unit excluding the first 8 octets
+                let seek_to = self_as_u8.offset(payload_offset as isize);
+
+                //lets get the first two indices which should be option_type and option_length fields
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                //Parse the option type field first then the option_length
+                let option_type: Option<Icmpv6OptionType> = FromPrimitive::from_u8(option_meta[0]);
+                let option_length = option_meta[1];
+                let option_length_octets = option_length * 8;
+
+                //Make sure the option is valid defined by the RFC. Note: a router can insert its own
```

Comment:
is there an auto formatter that can make sure that there is a space after the `//` and before the first word?

`//Make sure xxx` is painful, probably my programming upbringing

Reply:
I'll fix the comments. I tried to put enough detail in the comments in case it wasn't obvious what the parser does. however, all of the comments might be a little too much noise. if you feel its better for me to comment at the beginning of the method, i can do that